proto/p4/config/v1/p4types.proto

```diff
@@ -292,4 +292,211 @@ message P4NewTypeSpec {
   repeated StructuredAnnotation structured_annotations = 4;
 }
 
+//////// Generic Data types
+
+// Instead of duplicating the type spec for these
+// every time the type is used, we include the type spec once in this
+// GenericTypeInfo message and refer to the types by name in the
+// GenericDataTypeSpec message. We also
+// support annotations for these type specs which can be useful, e.g. to
+// identify well-known headers (such as ipv4).
+message GenericTypeInfo {
+  map<string, GenericStructTypeSpec> structs = 1;
+  map<string, GenericListTypeSpec> lists = 2;
+  map<string, GenericEnumTypeSpec> enums = 3;
+  map<string, GenericSerializableEnumTypeSpec> serializable_enums = 4;
+  map<string, GenericNewTypeSpec> new_types = 5;
+}
+
+// Describes Generic Data types.
+message GenericDataTypeSpec {
+  oneof type_spec {
+    GenericBitstringLikeTypeSpec bitstring = 1;
+    GenericFloatType float = 2;
+    GenericBoolType bool = 3;
+    GenericNamedType struct = 4;
+    GenericNamedType bag = 5;
+    GenericNamedType list = 6;
+    GenericNamedType set = 7;
+    GenericNamedType ordered_set = 8;
+    GenericNamedType string = 9;
+    GenericNamedType enum = 10;
+    GenericNamedType serializable_enum = 11;
+    GenericNamedType new_type = 12;
+    // P4 data type spec can help achieve parity with types being used
+    // in P4 externs. For example, if Register is using a type in P4
+    // Data type spec, then it will make it easier to map here when using
+    // GenericTables
+    P4DataTypeSpec p4_type = 13;
```

Comment:
If we keep this line in this definition, it would be a good idea to specify exactly when someone should use p4_type vs. bool, struct, enum, serializable_enum, or new_type above, so that among the two available choices, a consistent one is used.  Ideally the rule for which to use should be simple, e.g. "If the enum or serializable_enum is defined in the user's P4 program, use p4_type, otherwise use enum or serializable_enum instead of p4_type".

I have no idea what the rule should be to distinguish whether to use p4_type for `bool`, or use field `bool` directly above, though.

Reply:
Added a writeup